Source/WebCore/dom/CustomElementDefaultARIA.cpp

```diff
@@ -40,17 +41,101 @@ void CustomElementDefaultARIA::setValueForAttribute(const QualifiedName& name, c
     m_map.set(name, value);
 }
 
-const AtomString& CustomElementDefaultARIA::valueForAttribute(const QualifiedName& name) const
+const AtomString& CustomElementDefaultARIA::valueForAttribute(const Element& thisElement, const QualifiedName& name) const
 {
     auto it = m_map.find(name);
     if (it == m_map.end())
         return nullAtom();
-    return it->value;
+
+    const AtomString* result = &nullAtom();
+    std::visit(WTF::makeVisitor([&](const AtomString& stringValue) {
+        result = &stringValue;
+    }, [&](const WeakPtr<Element, WeakPtrImplWithEventTargetData>& weakElementValue) {
+        RefPtr elementValue = weakElementValue.get();
+        if (elementValue && (!elementValue->isConnected() || thisElement.isDescendantOrShadowDescendantOf(elementValue->rootNode())))
+            result = &elementValue->attributeWithoutSynchronization(HTMLNames::idAttr);
+    }, [&](const Vector<WeakPtr<Element, WeakPtrImplWithEventTargetData>>& elements) {
+        StringBuilder idList;
+        for (auto& weakElement : elements) {
+            RefPtr element = weakElement.get();
+            if (element && (!element->isConnected() || thisElement.isDescendantOrShadowDescendantOf(element->rootNode()))) {
+                if (idList.length())
+                    idList.append(' ');
+                idList.append(element->attributeWithoutSynchronization(HTMLNames::idAttr));
+            }
+        }
+    }), it->value);
+
+    return *result;
 }
 
 bool CustomElementDefaultARIA::hasAttribute(const QualifiedName& name) const
 {
     return m_map.find(name) != m_map.end();
 }
 
+Element* CustomElementDefaultARIA::elementForAttribute(const Element& thisElement, const QualifiedName& name) const
+{
+    auto it = m_map.find(name);
+    if (it == m_map.end())
+        return nullptr;
+
+    Element* result = nullptr;
+    std::visit(WTF::makeVisitor([&](const AtomString& stringValue) {
+        if (thisElement.isInTreeScope())
+            result = thisElement.treeScope().getElementById(stringValue);
+    }, [&](const WeakPtr<Element, WeakPtrImplWithEventTargetData>& weakElementValue) {
+        RefPtr elementValue = weakElementValue.get();
+        if (elementValue && (!elementValue->isConnected() || thisElement.isDescendantOrShadowDescendantOf(elementValue->rootNode())))
+            result = elementValue.get();
+    }, [&](const Vector<WeakPtr<Element, WeakPtrImplWithEventTargetData>>&) {
+        RELEASE_ASSERT_NOT_REACHED();
+    }), it->value);
+    return result;
+}
+
+void CustomElementDefaultARIA::setElementForAttribute(const QualifiedName& name, Element* element)
+{
```

Comment:
Nit: Add "ASSERT(element)"

Reply:
That's not true. This value can be null.

Reply:
Right, sorry for the noise.